fleetx/models/gpt_model/utils.py

```diff
@@ -0,0 +1,369 @@
+# Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
+# Copyright 2018 The OpenAI Team Authors and HuggingFace Inc. team.
+# Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import paddle
+from paddle import framework
+from paddle.nn import functional as F
+from paddle.autograd import PyLayer
+from paddle.fluid import core
+from paddle.fluid.dygraph.layers import Layer
+from paddle.distributed import fleet
+from paddle.distributed.fleet.base import topology as tp
+from paddle.distributed.fleet.meta_parallel import get_rng_state_tracker
+
+import numpy as np
+
+
+
+####################################################
+#                                                  #
+#        Distributed Communication Operator        #
+#                                                  #
+####################################################
+
+class ScatterOp(PyLayer):
+    # input shape: [s, b, h], n is mp parallelism
+    # after forward shape: [s/n, b, h]
+    @staticmethod
+    def forward(self, input):
+        hcg = fleet.get_hybrid_communicate_group()
+        group = hcg.get_model_parallel_group()
+        parallelism = group.nranks
+        rank = group.rank
+        assert input.shape[0] % parallelism == 0, \
+            "Input sequence length {0} can't be divided exactly \
+             by sequence parallelism {1}".format(input.shape[0], parallelism)
+        input = paddle.split(input, num_or_sections=parallelism, axis=0)[rank]
+        return input
+
+    @staticmethod
+    def backward(self, grad):
+        hcg = fleet.get_hybrid_communicate_group()
+        group = hcg.get_model_parallel_group()
+        parallelism = group.nranks
+        output_shape = grad.shape
+        output_shape[0] = output_shape[0] * parallelism
+        output = paddle.empty(shape=output_shape, dtype=grad.dtype)
+        group.process_group.all_gather(grad, output).wait()
+        return output
+
+class GatherOp(PyLayer):
```

Comment:
Duplicate codes with `ScatterOp`. Please follow the code suggestions in `ReduceScatterOp`.